File: src/domain/storage.js
const fs = require("fs");
const path = require("path");

const DATA_DIR = path.join(process.cwd(), "data");
const LAST_PLAN_FILE = path.join(DATA_DIR, "last-plan.json");
const PACKAGE_OVERRIDES_FILE = path.join(DATA_DIR, "package-overrides.json");

function ensureDataDir() {
  fs.mkdirSync(DATA_DIR, { recursive: true });
}

function readJson(filePath, fallback) {
  try {
    if (!fs.existsSync(filePath)) {
      return fallback;
    }

    const content = fs.readFileSync(filePath, "utf8");
    return JSON.parse(content);
  } catch (error) {
    return fallback;
  }
}

function writeJson(filePath, value) {
  ensureDataDir();
  fs.writeFileSync(filePath, JSON.stringify(value, null, 2));
}

function loadState() {
  return {
    lastPlan: readJson(LAST_PLAN_FILE, null),
    packageOverrides: readJson(PACKAGE_OVERRIDES_FILE, {})
  };
}

function savePlan(plan) {
  writeJson(LAST_PLAN_FILE, plan);
}

function savePackageOverrides(overrides) {
  writeJson(PACKAGE_OVERRIDES_FILE, overrides);
}

module.exports = {
  loadState,
  savePlan,
  savePackageOverrides
};

File: backend/storage.py
import json
from pathlib import Path


DATA_DIR = Path.cwd() / "data"
LAST_PLAN_FILE = DATA_DIR / "last-plan.json"
PACKAGE_OVERRIDES_FILE = DATA_DIR / "package-overrides.json"


def ensure_data_dir():
    DATA_DIR.mkdir(parents=True, exist_ok=True)


def read_json(file_path, fallback):
    try:
        if not file_path.exists():
            return fallback
        return json.loads(file_path.read_text(encoding="utf-8"))
    except Exception:
        return fallback


def write_json(file_path, value):
    ensure_data_dir()
    file_path.write_text(json.dumps(value, ensure_ascii=False, indent=2), encoding="utf-8")


def load_state():
    return {
        "lastPlan": read_json(LAST_PLAN_FILE, None),
        "packageOverrides": read_json(PACKAGE_OVERRIDES_FILE, {}),
    }


def save_plan(plan):
    write_json(LAST_PLAN_FILE, plan)


def save_package_overrides(overrides):
    write_json(PACKAGE_OVERRIDES_FILE, overrides)
